File: packages/flutter/lib/src/ui/hero_header.dart
/// The Home screen's tall branded header — background, avatar stack,
/// greeting and sub-greeting — and the rule by which it gets out of the way
/// once the visitor scrolls into the content below it.
///
/// Mirrors `ui/hero-header.ts`; see this file's "What this does not render"
/// section for what is deliberately left to the Home screen instead.
///
/// Two things live here, split the way the reference splits them: [HeroHeader]
/// is what the merchant configured, drawn, and is a plain [StatelessWidget]
/// that any test can pump on its own. [CollapsingHeroHeader] is the scroll
/// behaviour wrapped around it, and [heroCollapseDecision] is the rule that
/// behaviour obeys — a pure function over three numbers, so the oscillation
/// guard it exists for is assertable without a scroll gesture, a viewport or
/// a frame.
library;

import 'package:flutter/material.dart';

import '../config/remote_config.dart';
import '../theme/header_style.dart';
import 'image_safety.dart';

/// How many faces the row shows — `hero-header.ts`'s own `MAX_AVATARS`,
/// enforced again here rather than trusted: `header.avatars` is a plain
/// array on an opaque blob (remote_config.dart's header), and a fourth
/// avatar would overflow the row rather than being dropped the way an
/// older console version's own validation would have.
const int kMaxHeroAvatars = 3;

/// The Home screen's hero. Renders nothing at all — not an empty coloured
/// slab — when the merchant turned off every piece of it, matching
/// `hero-header.ts`'s own `data-empty` rule: "an empty hero is not a short
/// hero... a bare coloured slab above the transcript would read as a
/// rendering failure rather than as their choice."
///
/// ── What this does NOT render ───────────────────────────────────────────
///
/// `header.ctaEnabled` / `.ctaTitle`. In the JS widget, the hero carries its
/// own call-to-action because that widget has no separate Home screen to put
/// one on — "there is no conversation to start, so it focuses the composer"
/// (`hero-header.ts`'s own header). This package has a real Home screen with
/// its own "Send us a message" card (`home_screen.dart`, mirroring
/// `home-screen.ts`'s own separately-built `cta`), so the workaround this
/// hero would otherwise need does not apply — `ctaSubtitle` is read there
/// instead, unconditionally, matching `home-screen.ts`'s own unconditional
/// CTA card rather than this hero's gated one.
///
/// `aria-hidden` on the whole block. The JS hero is marked fully hidden from
/// assistive tech because every string in it is redundant with what the
/// composer right below it already conveys (its placeholder repeats the
/// greeting; its CTA's only action focuses that same composer, the next tab
/// stop regardless). Home has no composer sitting under this hero — the
/// greeting is the ONLY place that copy appears — so the redundancy the JS
/// choice relies on does not hold here, and hiding it would remove real
/// content from a screen-reader user instead of skipping a repeat.
class HeroHeader extends StatelessWidget {
  const HeroHeader({super.key, required this.config});

  final RemoteConfig config;

  @override
  Widget build(BuildContext context) {
    final HeaderAppearance header = config.header;
    final String? logoUrl = safeImageUrl(header.logoUrl ?? config.logoUrl);
    final bool showLogo = (header.showLogo ?? false) && logoUrl != null;

    final List<String> avatars = (header.showAvatars ?? false)
        ? (header.avatars ?? const <String>[])
            .map(safeImageUrl)
            .whereType<String>()
            .take(kMaxHeroAvatars)
            .toList(growable: false)
        : const <String>[];

    final String greeting = header.greeting ?? '';
    final String subGreeting = header.subGreeting ?? '';

    final bool isEmpty = !showLogo && avatars.isEmpty && greeting.isEmpty && subGreeting.isEmpty;
    if (isEmpty) return const SizedBox.shrink();

    final Color accent = Theme.of(context).colorScheme.primary;
    final Color backgroundColor = headerBackgroundColor(header, accent);
    final Color foregroundColor = readableOn(backgroundColor);
    final HeaderOverlay? overlay = headerOverlay(header);

    return Container(
      width: double.infinity,
      padding: const EdgeInsets.fromLTRB(20, 28, 20, 24),
      decoration: BoxDecoration(color: backgroundColor),
      child: Stack(
        children: <Widget>[
          if (overlay is HeaderImageOverlay) _HeroBackgroundImage(overlay: overlay),
          if (overlay is HeaderGradientOverlay) DecoratedBox(decoration: BoxDecoration(gradient: overlay.gradient)),
          Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              if (showLogo)
                Padding(
                  padding: const EdgeInsets.only(bottom: 12),
                  child: ClipRRect(
                    borderRadius: BorderRadius.circular(8),
                    child: Image.network(
                      logoUrl,
                      height: 32,
                      errorBuilder: (_, __, ___) => const SizedBox.shrink(),
                    ),
                  ),
                ),
              if (avatars.isNotEmpty)
                Padding(
                  padding: const EdgeInsets.only(bottom: 12),
                  child: _AvatarStack(avatars: avatars, showPresence: header.showPresence ?? false),
                ),
              if (greeting.isNotEmpty)
                Text(
                  greeting,
                  style: Theme.of(context)
                      .textTheme
                      .headlineSmall
                      ?.copyWith(color: foregroundColor, fontWeight: FontWeight.w600),
                ),
              if (subGreeting.isNotEmpty)
                Padding(
                  padding: const EdgeInsets.only(top: 4),
                  child: Text(
                    subGreeting,
                    style: Theme.of(context).textTheme.bodyMedium?.copyWith(color: foregroundColor.withOpacity(0.85)),
                  ),
                ),
            ],
          ),
        ],
      ),
    );
  }
}

/// How far past the top the visitor must scroll before the hero goes — and
/// the slack the collapse must SURVIVE in order to happen at all.
///
/// One number, used twice, because the two are the same fact: the distance
/// that has to exist on either side of the collapse for it not to undo
/// itself. The reference's `COLLAPSE_SLACK_PX` says the same thing.
const double kHeroCollapseSlackPx = 32;

/// What the hero should do about the scroll position it has just been told
/// about.
///
/// Three outcomes rather than a boolean, and the third one is the point:
/// refusing to collapse is NOT the same as expanding. See
/// [heroCollapseDecision].
enum HeroCollapseDecision {
  /// The visitor is at (or near) the top. The hero is shown in full.
  expand,

  /// The visitor has scrolled away and there is room to give the height back.
  collapse,

  /// The visitor has scrolled away, but collapsing would un-scroll the very
  /// scroll that caused it. Whatever the hero is doing now, it keeps doing.
  hold,
}

/// Whether the hero may collapse, given where the scroll is and how much
/// height collapsing would hand back.
///
/// ── The oscillation this closes ──────────────────────────────────────────
///
/// Collapsing returns the hero's full height to the scroll container. If the
/// content's current overflow is not comfortably larger than that height, the
/// framework clamps the scroll offset back toward the top the instant the
/// space arrives — which puts the hero back in its expand zone, which expands
/// it, which re-consumes the space, which scrolls it away again. The two
/// states re-trigger each other every frame, and what a customer sees is a
/// header strobing on a Home screen that barely overflows.
///
/// Two guards close it, and BOTH are needed:
///
///  1. **The slack margin.** [expand] until the visitor has scrolled more than
///     [kHeroCollapseSlackPx] — not the instant the offset leaves zero. This
///     alone makes the collapse unreachable when the whole overflow is smaller
///     than the slack, so the loop is never even woken for the sub-slack
///     scrolls that are the commonest way into it.
///  2. **The layout check, read AT COLLAPSE TIME.** Collapse only when, after
///     the height is handed back, the container will STILL be scrolled at
///     least that same slack. This covers the wider band of heights that full
///     removal exposes, which the margin alone cannot.
///
/// [maxScrollExtent] is Flutter's name for exactly what the reference computes
/// as `scrollHeight - clientHeight`: the content extent minus the viewport,
/// i.e. how far the container can scroll. So `maxScrollExtent - heroHeight` is
/// the reference's `overflow - freed`, and the guard is the same inequality.
///
/// [heroHeight] must be measured when this is called, never cached at build:
/// a config publish or an image finishing its load changes the hero's height,
/// and a stale number re-opens the loop this closes.
///
/// Refusing to collapse on a too-short Home is the CORRECT behaviour, not a
/// degraded one — there is nowhere for the freed space to go.
HeroCollapseDecision heroCollapseDecision({
  required double scrollOffset,
  required double maxScrollExtent,
  required double heroHeight,
}) {
  if (scrollOffset <= kHeroCollapseSlackPx) return HeroCollapseDecision.expand;
  if (maxScrollExtent - heroHeight <= kHeroCollapseSlackPx) {
    // NOT `expand`. Expanding here would be a second writer of the same state
    // fighting whatever put the hero where it is; "there is no room to do
    // this" is a refusal to act, not an instruction to act the other way.
    return HeroCollapseDecision.hold;
  }
  return HeroCollapseDecision.collapse;
}

/// A [HeroHeader] pinned above a scroll view, which gets out of the way once
/// the visitor scrolls into the content below it.
///
/// ── Why this owns the arrangement rather than just the header ────────────
///
/// It takes the scroll view as its [child] and builds the whole band-plus-
/// content column itself. That is not convenience: the hero has to sit ABOVE
/// the scroll view for a collapse to mean anything (a hero that were merely
/// the scroll view's first child would already be gone by the time the
/// visitor had scrolled past it), and a widget above a scroll view can reach
/// neither `Scrollable.of` — which searches ancestors, and the scroll view is
/// a sibling — nor that view's notifications, which bubble up past it. Owning
/// both halves puts this widget where the notifications actually pass and
/// spares every caller a [ScrollController] to thread through.
///
/// ── Collapsed means GONE, not a short bar ────────────────────────────────
///
/// There is no compact layer and no shrunken variant. Scrolled away from the
/// top the hero occupies NO space at all; scrolled back, it returns whole.
/// The reference is explicit that this is the behaviour, and the reason is
/// that a panel this size has room for one branded band — the app bar — and a
/// second, shorter one underneath it is two headers.
///
/// Spelled as [Align] with a zero `heightFactor` inside a [ClipRect], which is
/// the framework's own way of saying the reference's `height: 0;
/// overflow: hidden`. [Align] still lays its child out at full height and
/// merely reports zero for ITSELF, so the hero can be measured while
/// collapsed — which is what lets [heroCollapseDecision] read a real
/// [heroHeight] at the moment it decides, rather than the zero a removed
/// subtree would report.
///
/// ── A notification, not an observer, and why that is not a downgrade ─────
///
/// The reference picks an `IntersectionObserver` over a `scroll` listener
/// specifically because the DOM's scroll event misses anything that moves the
/// offset without a gesture — a programmatic `scrollTo`, a `scrollIntoView`
/// from elsewhere in the tree. Flutter has no such gap: `jumpTo`,
/// `animateTo` and `ensureVisible` all drive [ScrollPosition], and every one
/// of them emits a [ScrollUpdateNotification]. So the concern that chose the
/// observer there does not exist here, and the framework's own mechanism is
/// the right one.
///
/// A hero given a [child] that does not scroll simply never collapses, which
/// is correct rather than degraded — the same answer the reference gives an
/// environment with no `IntersectionObserver` at all.
class CollapsingHeroHeader extends StatefulWidget {
  const CollapsingHeroHeader({
    super.key,
    required this.config,
    required this.child,
  });

  final RemoteConfig config;

  /// The scrolling content the hero sits above. Given the remaining height.
  final Widget child;

  @override
  State<CollapsingHeroHeader> createState() => _CollapsingHeroHeaderState();
}

class _CollapsingHeroHeaderState extends State<CollapsingHeroHeader> {
  /// Measures the hero at its natural height, collapsed or not — see the
  /// class doc on why [Align] is what makes that possible.
  final GlobalKey _heroKey = GlobalKey();

  bool _collapsed = false;

  bool _onScroll(ScrollNotification notification) {
    // Only the view this widget is wrapped around, never one nested inside
    // its content: a horizontally scrolling row of chips down in the page
    // says nothing about how far Home itself has been scrolled.
    if (notification.depth != 0) return false;

    final RenderBox? hero =
        _heroKey.currentContext?.findRenderObject() as RenderBox?;
    if (hero == null || !hero.hasSize) return false;

    switch (heroCollapseDecision(
      scrollOffset: notification.metrics.pixels,
      maxScrollExtent: notification.metrics.maxScrollExtent,
      // Read HERE, at the moment of the decision, never cached at build.
      heroHeight: hero.size.height,
    )) {
      case HeroCollapseDecision.expand:
        _apply(false);
      case HeroCollapseDecision.collapse:
        _apply(true);
      case HeroCollapseDecision.hold:
        // Deliberately nothing. See `hold`'s own doc.
        break;
    }
    // Never swallowed: this only observes, and a pull-to-refresh or a scroll
    // metric watcher further up has as much right to hear about the scroll as
    // this does.
    return false;
  }

  void _apply(bool collapsed) {
    if (_collapsed == collapsed || !mounted) return;
    setState(() => _collapsed = collapsed);
  }

  @override
  Widget build(BuildContext context) {
    return NotificationListener<ScrollNotification>(
      onNotification: _onScroll,
      child: Column(
        children: <Widget>[
          ClipRect(
            child: Align(
              alignment: Alignment.topCenter,
              heightFactor: _collapsed ? 0.0 : 1.0,
              child: HeroHeader(key: _heroKey, config: widget.config),
            ),
          ),
          Expanded(child: widget.child),
        ],
      ),
    );
  }
}

class _HeroBackgroundImage extends StatelessWidget {
  const _HeroBackgroundImage({required this.overlay});

  final HeaderImageOverlay overlay;

  @override
  Widget build(BuildContext context) {
    return Positioned.fill(
      child: Stack(
        fit: StackFit.expand,
        children: <Widget>[
          Image.network(
            overlay.imageUrl,
            fit: BoxFit.cover,
            // Same graceful miss headerLayers gives a URL that fails to
            // decode — the base colour alone, never Flutter's red error box.
            errorBuilder: (_, __, ___) => const SizedBox.shrink(),
          ),
          DecoratedBox(decoration: BoxDecoration(color: Colors.black.withOpacity(overlay.scrimAlpha))),
        ],
      ),
    );
  }
}

class _AvatarStack extends StatelessWidget {
  const _AvatarStack({required this.avatars, required this.showPresence});

  final List<String> avatars;
  final bool showPresence;

  static const double _diameter = 36;
  static const double _overlap = 12;

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: _diameter,
      width: _diameter + (avatars.length - 1) * (_diameter - _overlap),
      child: Stack(
        children: <Widget>[
          for (int i = 0; i < avatars.length; i++)
            Positioned(
              left: i * (_diameter - _overlap),
              child: Stack(
                children: <Widget>[
                  Container(
                    width: _diameter,
                    height: _diameter,
                    decoration: BoxDecoration(
                      shape: BoxShape.circle,
                      border: Border.all(color: Colors.white, width: 2),
                      image: DecorationImage(
                        image: NetworkImage(avatars[i]),
                        fit: BoxFit.cover,
                        // A face that fails to load degrades to an empty
                        // circle rather than an unhandled image-stream
                        // exception — the same graceful miss every other
                        // merchant-supplied image in this widget gets.
                        onError: (_, __) {},
                      ),
                    ),
                  ),
                  // The presence dot rides the LAST face only — it says
                  // "someone is here", not "this particular person is", so
                  // one is the honest count regardless of how many faces are
                  // shown. Mirrors hero-header.ts's own placement exactly.
                  if (showPresence && i == avatars.length - 1)
                    Positioned(
                      right: 0,
                      bottom: 0,
                      child: Container(
                        width: 10,
                        height: 10,
                        decoration: BoxDecoration(
                          shape: BoxShape.circle,
                          color: Colors.green,
                          border: Border.all(color: Colors.white, width: 2),
                        ),
                      ),
                    ),
                ],
              ),
            ),
        ],
      ),
    );
  }
}

File: packages/widget/src/ui/hero-header.ts
// The tall branded header — the console's `design: 'hero'`.
//
// ── What "hero" means HERE, which is not quite what it means in the console ─
//
// In `chatsupport_react` the hero is a HOME SCREEN: a tall panel with a
// greeting and a "send us a message" card, and picking that card is what opens
// a conversation. It collapses once the visitor scrolls into one.
//
// This widget NOW has that home/conversation split (`ui/screens.ts`,
// `ui/home-screen.ts`), and the hero is mounted as the Home screen's own
// banner — `widget.ts` shows it exactly while `screens.current() === 'home'`
// and `design === 'hero'`. Before that wiring landed, this widget's panel WAS
// the conversation, and the hero stood in for a home screen by keying off
// "the transcript is EMPTY" instead; that is no longer how any of it works,
// but it is worth naming here because the trigger below reads similarly and
// is a different thing:  `.dh-hero[data-empty="true"]` — "this hero has
// nothing to draw" — is unrelated to which SCREEN is showing.
//
// ── The hero has no call to action, and this is decided ───────────────────
//
// It used to build a `.dh-hero-cta` from `header.ctaEnabled`/`ctaTitle`/
// `ctaSubtitle`, calling back into the same "open the new-conversation
// surface" flow that Home's own card and the Messages screen's "New
// conversation" button open. This file used to describe that overlap as an
// open console/design question. It is not open any more: the answer is that
// the hero renders no CTA at all.
//
// What settled it is that the overlap is not conditional. The hero shows only
// while the Home screen is showing — `widget.ts`'s `syncScreens` gates it on
// `onHome && design === 'hero'`, a strict subset of the times Home is up — and
// `ui/home-screen.ts` builds `.dh-home-cta` on Home unconditionally, so an
// enabled hero CTA is never an alternative to Home's card, it is always a
// SECOND card, directly above it, making the identical offer and leading to
// the identical destination. Two of them is one more than the screen is
// asking for, and the hero's was the weaker of the two: it is inside this
// block's `aria-hidden`, so it carried `tabindex="-1"`, and `ui/focus.ts`'s
// `isHidden` walks up to that same `aria-hidden` and kept it out of the focus
// trap as well. No keyboard user and no screen reader could reach it.
//
// It was still PAINTED and still clickable, so a sighted pointer user does
// lose an affordance here — that is the honest cost and it is not nothing.
// What no user loses is a DESTINATION: Home's own card sits one gap below,
// makes the identical offer and opens the identical flow.
//
// And the removal is a REPAIR, not only a tidy-up. A control the widget drew,
// styled as a button and invited a click on, that keyboard and screen-reader
// users could not operate at all, is a defect on its own terms — it had been
// shipping as one. Deleting it closes that, and leaves one affordance on Home
// which every user can actually reach.
//
// The three console fields still parse and still round-trip (`src/config.ts`,
// `src/remote-config.ts` — they are the published wire schema). They simply
// no longer reach a rendered button; see `heroContentFrom` at the bottom of
// this file for what a merchant who set them sees instead.
//
// ── The collapse: expanded or gone, nothing in between ─────────────────────
//
// The first cut of this feature kept an Intercom-style 66px compact bar
// (logo + faces) pinned behind the tall hero. The product owner rejected it
// against a live tenant — on a hero with no logo and no faces the bar is an
// EMPTY 66px strip of dead colour covering content — and re-specced the
// behaviour: scrolled away from the top, the hero must occupy NO space at
// all, so "Recent conversation"/"Common Questions" slide up directly under
// the panel's standard header row; scrolled back to the top, it returns.
// There is no compact layer any more. One state machine: expanded, or gone.
//
// `.dh-hero` is a `flex: none` band ABOVE `ui/home-screen.ts`'s own scroll
// container (`.dh-home`, `overflow-y: auto`) — nothing about the hero itself
// ever scrolls, so "gone" is a straight `height: 0; overflow: hidden` snap on
// `[data-collapsed="true"]` (ui/styles.ts). Snapping the height rather than
// animating it is what actually returns the freed space to `.dh-home` (a
// `flex: none` sibling shrinking is the only way its `flex: 1` neighbour gets
// more room) without a second, competing animation. The content's own 180ms
// opacity fade (`.dh-hero-full` in ui/styles.ts) is the only thing that
// animates, and it matters on the way BACK: the band's height returns in one
// frame and the content fades in over it instead of popping. A snap is not
// motion in the vestibular sense a slide or a transform is, so it gets no
// `prefers-reduced-motion` carve-out — same stance as `.dh-msg-reply`.
//
// ── The trigger: a zero-space marker at the top of `.dh-home` ──────────────
//
// `watchScroll(scrollHost)` is how `widget.ts` tells this component where
// "the top of Home's content" actually is — `ui/home-screen.ts` owns
// `.dh-home`'s markup and this file has no reason to reach past that
// boundary, so rather than have `home-screen.ts` grow a bespoke sentinel slot
// for one caller, this inserts its own marker into the element it is handed,
// the same "caller supplies the mount point, callee owns what goes in it"
// shape `widget.ts` already uses for `homeQuestionsSlot`.
//
// The marker (`.dh-hero-sentinel`) is `position: absolute; height: 1px`, not
// a normal flow child — `.dh-home` lays its children out with `gap`, and gap
// applies to every flex item regardless of that item's own margins, so a real
// flow child would permanently widen the space above the CTA card by one
// gap's worth for a pixel nobody is meant to see. Absolute costs nothing.
//
// An `IntersectionObserver` rooted at `scrollHost` watching that marker is
// simpler and cheaper than a `scroll` listener doing the same arithmetic on
// every frame: the browser only calls back on the two edges this component
// actually cares about (crossed into view / crossed out of view), and it
// keeps working correctly through anything that moves `.dh-home`'s scroll
// position without firing a `scroll` event on it directly (programmatic
// `scrollTo`, `scrollIntoView` from elsewhere in the tree).
//
// ── The oscillation trap, which full removal makes WORSE, not better ───────
//
// Collapsing hands the hero's height back to `.dh-home`, which can un-scroll
// the very scroll that caused the collapse: the container grows, the browser
// clamps `scrollTop` back toward 0, the marker re-enters view, the hero
// re-expands, the overflow comes back, and it repeats. A real-Chrome run of
// the 66px version surfaced exactly this loop on a near-empty Home, and the
// fix was a positive `rootMargin` (below). Removing the hero ENTIRELY frees
// its full height rather than height-minus-66px, so the band of content
// heights that can bounce is wider now — the static 32px margin alone no
// longer covers it, and `watchScroll` adds a layout check at collapse time:
// it refuses to collapse unless the container will STILL be scrolled past
// the margin after the hero's height is handed back. See the guard inline.

import type { HeaderAppearance } from '../config.js';

import { DEFAULT_AGENT_AVATARS, DEFAULT_AVATAR_IMAGE, DEFAULT_LOGO_IMAGE, el, safeImageUrl } from './dom.js';

/**
 * A logo/avatar `<img>` that swaps to `fallback` instead of sitting as a
 * broken-image glyph when the browser cannot actually load `src` — see
 * DEFAULT_AVATAR_IMAGE/DEFAULT_LOGO_IMAGE's own docs in dom.ts for why this
 * can happen even though `src` already passed `safeImageUrl`. Callers pass
 * the agent silhouette for a person's photo and the Dhaam AI wordmark for a
 * brand logo — the two 404 cases read very differently.
 */
function imgWithFallback(
  attrs: Record<string, string | number | boolean | null | undefined>,
  fallback: string = DEFAULT_AVATAR_IMAGE,
): HTMLImageElement {
  const img = el('img', { attrs });
  img.addEventListener('error', () => { img.src = fallback; }, { once: true });
  return img;
}

export interface HeroContent {
  readonly showLogo: boolean;
  readonly logoUrl: string;
  readonly showAvatars: boolean;
  readonly showPresence: boolean;
  readonly avatars: readonly string[];
  readonly greeting: string;
  readonly subGreeting: string;
}

export interface HeroHeaderView {
  readonly node: HTMLElement;
  /** Avatars displayed inside the top header row when the hero is collapsed on scroll. */
  readonly headerAvatars: HTMLElement;
  /** Rebuilds the hero from a new appearance. Cheap; runs at most once per publish. */
  render(content: HeroContent): void;
  /**
   * Starts removing this hero (height 0, nothing left behind) once
   * `scrollHost` — Home's own scroll container (`ui/home-screen.ts`'s
   * `.dh-home`) — has been scrolled away from the top, and restoring it once
   * the visitor returns there. See the module header for why this, and not a
   * `scroll` listener, is what drives it.
   *
   * Meant to be called once, after the Home screen exists. A second call is
   * a no-op rather than a second observer — nothing in `widget.ts` recreates
   * the Home screen mid-lifetime, so stacking observers would only ever be a
   * bug, never a legitimate re-target.
   */
  watchScroll(scrollHost: HTMLElement): void;
  /** Disconnects the scroll watch and removes its marker. Safe to call twice. */
  destroy(): void;
}

/**
 * How many faces the row shows.
 *
 * Three, the console's own cap, enforced again here rather than trusted: the
 * field is a plain array on an opaque blob, and a fourth avatar would overflow
 * the row rather than being ignored the way an older console version's
 * validation would have.
 */
const MAX_AVATARS = 3;

/**
 * How far past the sentinel the visitor must scroll before the hero goes —
 * the `rootMargin` on the observer below, and the slack the collapse guard
 * requires to SURVIVE the collapse. One number, used twice, because the two
 * checks describe the same edge: "is the marker meaningfully out of view".
 */
const COLLAPSE_SLACK_PX = 32;

/**
 * The face row.
 *
 * Takes already-filtered, already-capped URLs rather than the raw
 * `HeroContent.avatars` — `render()` below does that filtering exactly once
 * per publish, so which faces survived the allowlist (`safeImageUrl`) and the
 * cap (`MAX_AVATARS`) is decided in one place.
 */
function buildAvatarRow(faces: readonly string[], showPresence: boolean, isHeader = false): HTMLElement | null {
  if (faces.length === 0) return null;
  const avatarClass = isHeader ? 'dh-header-hero-avatar' : 'dh-hero-avatar';
  return el('div', {
    attrs: { class: isHeader ? 'dh-header-hero-avatars-row' : 'dh-hero-avatars' },
    children: faces.map((src, index) => {
      const fallback = DEFAULT_AGENT_AVATARS[index % DEFAULT_AGENT_AVATARS.length] || DEFAULT_AVATAR_IMAGE;
      const effectiveSrc = (!src || src.includes('/assets/chat/agent-')) ? fallback : src;
      return el('span', {
        attrs: { class: avatarClass },
        children: [
          imgWithFallback({ src: effectiveSrc, alt: '' }, fallback),
          // The presence dot rides the LAST face only — it says "someone
          // is here", not "this particular person is", so one is the
          // honest number regardless of how many faces are shown.
          ...(showPresence && index === faces.length - 1
            ? [el('span', { attrs: { class: isHeader ? 'dh-header-hero-presence' : 'dh-hero-presence' } })]
            : []),
        ],
      });
    }),
  });
}

// Takes no callbacks. It had exactly one — the CTA's — and the CTA is gone
// (see the module header); everything left in this block is decoration that
// nothing can press.
export function createHeroHeader(): HeroHeaderView {
  const full = el('div', { attrs: { class: 'dh-hero-full' } });
  const headerAvatars = el('div', { attrs: { class: 'dh-header-hero-avatars', 'aria-hidden': 'true' } });

  // `aria-hidden` on the whole block, and this is deliberate rather than
  // careless. Every string in it is decoration that the panel already conveys:
  // the greeting repeats what the composer's placeholder asks for, and the
  // logo and faces are branding. Announcing all of it before the conversation
  // would put several lines of marketing in front of a screen reader user
  // every time they open the chat. Nothing in here is interactive any more
  // either — the CTA was the only control this block ever had, and the reason
  // it needed an explicit `tabindex="-1"` (a focusable control inside an
  // aria-hidden subtree is the one combination that genuinely breaks
  // assistive tech). With it gone, the promise this attribute makes is
  // structurally true rather than maintained by hand.
  const node = el('div', {
    attrs: { class: 'dh-hero', hidden: true, 'aria-hidden': 'true' },
    children: [full],
  });

  function render(content: HeroContent): void {
    const logo = content.showLogo ? safeImageUrl(content.logoUrl) : null;
    const rawFaces = content.showAvatars
      ? content.avatars
          .map((url) => safeImageUrl(url))
          .filter((url): url is string => url !== null)
          .slice(0, MAX_AVATARS)
      : [];
    const faces = content.showAvatars && rawFaces.length === 0 ? DEFAULT_AGENT_AVATARS : rawFaces;

    const fullChildren: Node[] = [];

    if (logo !== null) {
      fullChildren.push(imgWithFallback({ class: 'dh-hero-logo', src: logo, alt: '' }, DEFAULT_LOGO_IMAGE));
    }

    const avatars = buildAvatarRow(faces, content.showPresence);
    if (avatars !== null) fullChildren.push(avatars);

    // Also populate the compact header avatars row for collapsed state
    const headerAvatarRow = buildAvatarRow(faces, content.showPresence, true);
    headerAvatars.replaceChildren(...(headerAvatarRow !== null ? [headerAvatarRow] : []));

    if (content.greeting !== '') {
      fullChildren.push(el('p', { attrs: { class: 'dh-hero-greeting' }, text: content.greeting }));
    }
    if (content.subGreeting !== '') {
      fullChildren.push(el('p', { attrs: { class: 'dh-hero-sub' }, text: content.subGreeting }));
    }

    // No CTA. `header.ctaEnabled`/`ctaTitle`/`ctaSubtitle` are not read here
    // at all — see the module header for why Home's card is the one that
    // survived, and `heroContentFrom` for what those fields still do.

    full.replaceChildren(...fullChildren);

    // An empty hero is not a short hero — a merchant who turned off every
    // piece of it has turned the block off, and a bare coloured slab above the
    // transcript would read as a rendering failure rather than as their
    // choice. The caller's own visibility rule still applies on top of this.
    node.dataset['empty'] = fullChildren.length === 0 ? 'true' : 'false';
  }

  let observer: IntersectionObserver | null = null;
  let sentinel: HTMLElement | null = null;

  function watchScroll(scrollHost: HTMLElement): void {
    if (observer !== null) return;
    // No polyfill, no crash — the same "feature-detect a browser global"
    // shape `ui/chime.ts` uses for `AudioContext`. Every browser this widget
    // otherwise supports has had `IntersectionObserver` for years, so this is
    // not really about legacy support; it is about every environment that
    // constructs a widget WITHOUT being a real browser tab — a test harness,
    // a server-side render pass a host wraps this in — none of which has any
    // scrolling for this feature to watch anyway. Collapsing never happening
    // there is correct, not degraded: the merchant's hero still renders,
    // expanded, exactly as it did before this feature existed.
    if (typeof IntersectionObserver === 'undefined') return;

    sentinel = el('div', { attrs: { class: 'dh-hero-sentinel', 'aria-hidden': 'true' } });
    scrollHost.prepend(sentinel);

    observer = new IntersectionObserver(
      (entries) => {
        const entry = entries[entries.length - 1];
        if (entry === undefined) return;
        // Intersecting the scroll host's own viewport = at (or near) the
        // top = expanded. Anything else = the visitor has scrolled away.
        if (entry.isIntersecting) {
          node.dataset['collapsed'] = 'false';
          return;
        }
        // ── The collapse guard — see the module header's oscillation note ──
        //
        // Removing the hero hands its FULL height back to `.dh-home`. If the
        // content's current overflow is not comfortably larger than that, the
        // browser clamps `scrollTop` back into the marker's expand zone the
        // instant the space arrives, and the two states re-trigger each other
        // every frame. So: collapse only when, AFTER the hero's height is
        // returned, the container will still be scrolled at least the same
        // slack past the marker that `rootMargin` demanded to get here.
        // Refusing to collapse on a too-short Home is the correct behaviour,
        // not a degraded one — there is nowhere for the freed space to go.
        //
        // `offsetHeight` is read here, at the moment of the decision, rather
        // than cached at render: config publishes and image loads change the
        // hero's height, and a stale number re-opens the loop this closes.
        const freed = node.offsetHeight;
        const overflow = scrollHost.scrollHeight - scrollHost.clientHeight;
        if (overflow - freed <= COLLAPSE_SLACK_PX) return;
        node.dataset['collapsed'] = 'true';
      },
      // `root`, not the default (the browser viewport) — `.dh-home` scrolls
      // internally inside a fixed-size panel; the panel itself never moves
      // relative to the page, so watching the page's viewport would never
      // fire at all. `threshold: 0` (the default, spelled out because a
      // future edit reaching for "a little debounce" should have to change
      // an explicit value rather than discover an implicit one): the marker
      // is 1px tall, so "any of it visible" and "all of it visible" are the
      // same question anyway.
      //
      // `rootMargin`'s top value is NOT decorative — a real-Chrome run of
      // the 66px-bar version of this code surfaced the feedback loop the
      // module header describes on a Home that only overflowed by a few
      // pixels. A POSITIVE top margin grows the effective root upward, so
      // the marker still counts as "in view" (expanded) until the visitor
      // has scrolled more than 32px past it — not the instant it crosses 0 —
      // which makes the collapse unreachable when the whole overflow is
      // smaller than the slack. The layout guard in the callback above
      // handles the wider band of heights that full removal exposes; this
      // margin still earns its place by never waking the callback at all
      // for sub-slack scrolls.
      {
        root: scrollHost,
        threshold: 0,
        rootMargin: `${COLLAPSE_SLACK_PX}px 0px 0px 0px`,
      },
    );
    observer.observe(sentinel);
  }

  function destroy(): void {
    observer?.disconnect();
    observer = null;
    sentinel?.remove();
    sentinel = null;
  }

  return { node, headerAvatars, render, watchScroll, destroy };
}

/**
 * The hero's content, assembled from the two places the console keeps it.
 *
 * `header.logoUrl` is the hero's own logo; `appearance.logoUrl` is the brand
 * mark the console writes when `avatarMode` is `'logo'`. A merchant who
 * uploaded one and not the other means the same thing by either, so the
 * header's own wins and the top-level one stands behind it — which is the
 * difference between `showLogo: true` rendering something and rendering
 * nothing at all.
 *
 * ── Where `ctaEnabled` / `ctaTitle` / `ctaSubtitle` stop ──────────────────
 *
 * Here. They are on `HeaderAppearance` and they stay there — `src/config.ts`
 * and `src/remote-config.ts` are the published wire schema, a console that
 * writes them must keep round-tripping, and dropping a field a merchant has
 * saved is a different and much larger decision than not drawing a button.
 * They are simply never copied onto `HeroContent`, so nothing downstream can
 * read them.
 *
 * What a merchant who set them sees: the hero draws their logo, faces,
 * greeting and sub-line, and no button — and Home's own card, one gap below
 * where theirs would have been, carries the "send us a message" offer with
 * ITS OWN copy (`ui/home-screen.ts`'s `ctaCopy`), not the `ctaTitle` they
 * typed. So a merchant who wrote a custom hero CTA title loses that wording;
 * the affordance itself is not lost, and their title never appears twice.
 */
export function heroContentFrom(
  header: HeaderAppearance,
  fallbackLogoUrl: string,
): HeroContent {
  const avatars = (header.avatars && header.avatars.length > 0)
    ? header.avatars
    : DEFAULT_AGENT_AVATARS;
  return {
    showLogo: header.showLogo,
    logoUrl: header.logoUrl.trim() === '' ? fallbackLogoUrl : header.logoUrl,
    showAvatars: header.showAvatars,
    showPresence: header.showPresence,
    avatars,
    greeting: header.greeting,
    subGreeting: header.subGreeting,
  };
}
